spec/multi_patient_api/bulk_data_group_export_validation_spec.rb

```diff
@@ -0,0 +1,514 @@
+
+      allow_any_instance_of(runnable).to receive(:resource_is_valid?).and_return(true)
+      result = run(runnable, patient_input)
+
+      expect(result.result).to eq('pass')
+    end
+  end
+
+  describe '[Group export has at least two patients]' do
+    let(:runnable) { group.tests[3] }
+
+    it 'skips when no patient ids have been stored' do
+      result = run(runnable)
+
+      expect(result.result).to eq('skip')
+      expect(result.result_message).to eq('No Patient resources processed from bulk data export.')
+    end
+
+    it 'fails when less than two patient ids are stored' do
+      scratch[:patient_ids_seen] = ['one_id']
+      result = run(runnable)
+
+      expect(result.result).to eq('fail')
+      expect(result.result_message).to eq('Bulk data export did not have multiple Patient resources.')
+    end
+
+    it 'passes when two or more patient ids are stored' do
+      scratch[:patient_ids_seen] = ['one_id', 'two_id']
+      result = run(runnable)
+
+      expect(result.result).to eq('pass')
+    end
+  end
+
+  describe '[Patient IDs match those expected in Group]' do
+    let(:runnable) { group.tests[4] }
+    let(:input) { { bulk_patient_ids_in_group: 'one_id, two_id, three_id' } }
+
+    it 'omits when no patient ids have been stored' do
+      result = run(runnable)
+
+      expect(result.result).to eq('omit')
+      expect(result.result_message).to eq('No patient ids were given.')
+    end
+
+    it 'fails when the input patient ids do not match those stored' do
+      scratch[:patient_ids_seen] = ['one_id', 'one_id', 'one_id']
+      result = run(runnable, input)
+
+      expect(result.result).to eq('fail')
+      expect(result.result_message).to eq('Mismatch between patient ids seen (one_id, one_id, one_id) and patient ids expected (one_id, two_id, three_id)')
+    end
+
+    it 'fails when more input patient ids than ids stored' do
+      scratch[:patient_ids_seen] = ['one_id', 'two_id']
+      result = run(runnable, input)
+
+      expect(result.result).to eq('fail')
+      expect(result.result_message).to eq('Mismatch between patient ids seen (one_id, two_id) and patient ids expected (one_id, two_id, three_id)')
+    end
+
+    it 'fails when less input patient ids than ids stored' do
+      scratch[:patient_ids_seen] = ['one_id', 'two_id', 'three_id', 'four_id']
+      result = run(runnable, input)
+
+      expect(result.result).to eq('fail')
+      expect(result.result_message).to eq('Mismatch between patient ids seen (one_id, two_id, three_id, four_id) and patient ids expected (one_id, two_id, three_id)')
+    end
+
+    it 'passes when the input patient ids do not match those stored' do
+      scratch[:patient_ids_seen] = ['one_id', 'two_id', 'three_id']
+      result = run(runnable, input)
+
+      expect(result.result).to eq('pass')
+    end
+  end
+
+  describe '[AllergyIntolerance resources returned conform to the US Core AllergyIntolerance Profile] test' do
```

Comment:
Why are AllergyIntolerance, CarePlan, DiagnosticReport, and Observation specifically tested? Should more tests be done on the tester class instead?

Reply:
Short answer: yes, I need to write more tests for the **tester** class. For the bulk_data_group_export_validation tests though, I think these are good. AllergyIntolerance and CarePlan just came up first and I think those tests cover most single-profile resource testing. I did DiagnosticReport and Observation tests specifically because those are the multiple profile cases, so I figured it'd be worth going deeper on those. Overall, it's my understanding that most single-profile resources like AllergyIntolerance and CarePlan will have the same behavior as what is tested  - we don't do any special profile guessing, so I'm not sure what would be significantly different. Feel free to disagree. Though, on second thought, I'm going to add some device tests as well.